### PR TITLE
Reduce memory footprint for dynamic buckets by sharing common values …

### PR DESCRIPTION
**Pull request by maniksurtani**
…across all dynamic buckets created from the same template

### CODE REVIEW

File: buckets/redis/bucket.go [hunk 1]
```diff
@@ -26,26 +26,39 @@ const (
...
+	cfg       *pbconfig.BucketConfig
+	factory   *bucketFactory
+	redisKeys []string // {tokensNextAvailableRedisKey, accumulatedTokensRedisKey}
 }
 
 type bucketFactory struct {
+	// Embedded mutex
+	sync.Mutex
+
+	// Ref counts of redisBucketCommon instances for each namespace. Protected by embedded mutex.
```
Comment (kkaneda) [inline buckets/redis/bucket.go]: might want to mention that this is only for dynamic buckets?

File: buckets/redis/bucket.go [hunk 2]
```diff
@@ -113,19 +128,39 @@ func (bf *bucketFactory) NewBucket(namespace, bucketName string, cfg *pbconfig.B
...
 	}
 
-	rb := &redisBucket{
-		dyn,
+	var rbc *redisBucketCommon
+
+	if dyn {
+		var exists bool
+		bf.Lock()
+		defer bf.Unlock()
```
Comment (kkaneda) [inline buckets/redis/bucket.go]: we need to surround this in a function or just call `bf.Unlock` at the end of the`if` block..

https://play.golang.org/p/-MBcYET03D

Comment (maniksurtani) [inline buckets/redis/bucket.go]: Yes, the intention is for `bf.Unlock()` to be called when the function returns.

Comment (kkaneda) [inline buckets/redis/bucket.go]: oh i see. sg!

**Comment by maniksurtani** [comment]
@kkaneda updated + added tests

**Pull request merged**